File: Data Structures & Algorithms/edit-distance/submission-1.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int rows = word1.size() + 1, cols = word2.size() + 1;
        vector<vector<int>> dp(rows, vector<int>(cols, 0));
        
        for (int c = 0; c < cols - 1; c++) dp[rows - 1][c] = cols - 1 - c; 
        for (int r = 0; r < rows - 1; r++) dp[r][cols - 1] = rows - 1 - r;

        // for (int i = 0; i < dp.size(); i++) {
        // for (int j = 0; j < dp[i].size(); j++) {
        //     cout << dp[i][j] << " ";
        //     }
        // cout << endl;
        // }
        for (int r = rows - 2; r >= 0; r--){
            for (int c = cols - 2; c >= 0; c--){
                if (word1[r] == word2[c]) dp[r][c] = dp[r + 1][c + 1];
                else dp[r][c] = 1 + min({dp[r + 1][c], dp[r][c + 1], dp[r + 1][c + 1]});
            }
        }
        return dp[0][0];
    }
};

File: Data Structures & Algorithms/islands-and-treasure/submission-1.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        ROWS, COLS = len(grid), len(grid[0])
        dirs = [[-1, 0], [1, 0], [0, 1], [0, -1]]
        q = deque([])
        visited = set()

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 0:
                    q.append([r, c])
                    visited.add((r, c))
                elif grid[r][c] == -1:
                    visited.add((r, c))
        level = 0
        while q:
            level += 1
            for _ in range(len(q)):
                r, c = q.popleft()
                for dr, dc in dirs:
                    new_r, new_c = r + dr, c + dc
                    if (0 <= new_r < ROWS and 
                        0 <= new_c < COLS and
                        (new_r, new_c) not in visited):
                        grid[new_r][new_c] = level
                        visited.add((new_r, new_c))
                        q.append([new_r, new_c])
